### PR TITLE
Types for meteor/dburles:collection-helpers [meteor-dburles-collection-helpers]

### PR DESCRIPTION
**Pull request by artemiswkearney**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/meteor-dburles-collection-helpers/mongo.d.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                },
+            ): (Full<T> & T) | undefined;
+            // ditto
+            // tslint:disable-next-line ban-types
+            insert(doc: OptionalId<Data<T>>, callback?: Function): string;
+            update(selector: Selector<T> | ObjectID | string, modifier: Modifier<Data<T>>, options?: {
+                multi?: boolean;
+                upsert?: boolean;
+                arrayFilters?: Array<{ [identifier: string]: any }>;
+				// ditto
```
Comment (danger-public) [inline types/meteor-dburles-collection-helpers/mongo.d.ts:67]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by artemiswkearney** [comment]
Known issue I'm giving up on for the moment:
`
const upsert = <T extends { _id?: string }>(collection: Mongo.Collection<T>, id: string, value: Mongo.OptionalId<Data<T>>) => collection.upsert(id, { ...value, _id: id });
`
does not compile, while
`
const upsert = <T extends { _id?: string }>(collection: Mongo.Collection<T>, id: string, value: Mongo.OptionalId<T>) => collection.upsert(id, { ...value, _id: id });
`
does.
The latest commit got this working for upsert on collections of known types, but not in the generic case. If anyone knows how to fix it in the generic case, I'd greatly appreciate help.

**Comment by artemiswkearney** [comment]
... problems run deeper than I thought; having trouble with `T | null` properties now
probably don't merge this until I figure out what's wrong there

**Comment by artemiswkearney** [comment]
Fixed the problems with null, and improved a bunch of other stuff.

**Review by andrewbranch — APPROVED** [review]

**Comment by artemiswkearney** [comment]
Ready to merge

**Pull request merged**